3_extracting_features/DPutils.py

```diff
@@ -0,0 +1,220 @@
+            # Count is set to one to only change the first instance that '_1_' occurs (only change channel not site)
+            channel_path = pathlib.Path(str(image_paths).replace('_1_', f'_{index+1}_'), count=1)
+
+            # Since the channel path will keep 'DAPI' for all images, these if statements will change the name to respective channel based on name
+            if '01_2_' in str(channel_path):
+                channel_path = pathlib.Path(str(channel_path).replace('DAPI', 'GFP'))
+            if '01_3_' in str(channel_path):
+                channel_path = pathlib.Path(str(channel_path).replace('DAPI', 'RFP'))
+            # Save paths to each image into the index.csv under the correct column (within dictionary) with the name of the channel
+            file_data[channel]= os.path.relpath(channel_path, DP_images_path)
+        
+        # Add Genotype and Genotype_Replicate (hard coded to 1 for NF1 data) with to file_data dictionary
+        file_data["Genotype"] = genotype
+        file_data["Genotype_Replicate"] = 1
+
+        # Append all of the data into an index.csv file
+        index_csv_data.append(file_data)
+
+    return pd.DataFrame(index_csv_data)
+
+def compile_training_locations(
+    index_csv_path: pathlib.Path,
+    segmentation_data_path: pathlib.Path,
+    save_path: pathlib.Path,
+    object: str,
+):
+    """Compile well-site-object.csv file with cell locations, saving to save_path/plate/well
+    Args:
+        index_csv_path (pathlib.Path): Path to index.csv file for object (nuc or cyto) DeepProfiler project
+        segmentation_data_path (pathlib.Path): Path to segmentation folder with .tsv locations files
+        save_path (pathlib.Path): Path to save location files
+        object (str): Object to find segmentation locations for, either "nuc" or "cyto"
+    """
+    # Reads in csv and iterate through the rows from the plate, well and site columns
+    index_csv = pd.read_csv(index_csv_path)
+    for index, row in index_csv.iterrows():
+        plate = row["Metadata_Plate"]
+        well = row["Metadata_Well"]
+        site = row["Metadata_Site"]
+
+        # Gets identifier string that matches identifier from segmented images tsvs
+        identifier_details = row["DNA"].split("/")[-1].split("_")[0:4]
+        identifier_well = identifier_details[0]
+        identifier_site = identifier_details[3]
+        identifier = f"{identifier_well}_{identifier_site}"
+
+        # Note: The name is hard coded to "Nuclei.csv" due to how the pretrained model works. Even though it should be Cytoplasm for the cyto_project,
+        # it has to be named "Nuclei" to prevent errors when running DeepProfiler (e.g saying no cells to profile when there are)
+        locations_save_path = pathlib.Path(
+            f"{save_path}/{plate}/{well}-{site}-Nuclei.csv"
+        )
+
+        # Skips a field if the locations have already been found
+        if locations_save_path.is_file():
```

Comment:
nice!

I wonder... do these ever have to be rewritten? if so, you could add a `force=False` argument to the function and, if `force=True`, then you don't check `if locations_save_path.is_file()`

Reply:
(Note, this isn't me requesting this enhancement, this is meant for discussion)

Reply:
@gwaybio I mean if it isn't on, that means it would just overwrite the files that are already there right? 

I have noticed that if anything goes wrong after compiling the project, you need to delete the `input` and `output` folders before rerunning because errors happen if the directory already exists. Likely something I can fix but if anyone needs to rerun the function, they delete all the directories anyway. 

This deserves a deeper discussion but I think I will keep this function for now.

Reply:
sounds good! Thanks for the discussion and for making an executive decision 👍